src/stan/model/indexing/rvalue.hpp

```diff
@@ -151,16 +151,17 @@ inline auto rvalue(Vec&& v, const char* name, index_uni idx) {
  * the indexed size.
  */
 template <typename EigVec, require_eigen_vector_t<EigVec>* = nullptr>
-inline plain_type_t<EigVec> rvalue(EigVec&& v, const char* name,
-                                   const index_multi& idx) {
-  const auto v_size = v.size();
-  const auto& v_ref = stan::math::to_ref(v);
-  plain_type_t<EigVec> ret_v(idx.ns_.size());
-  for (int i = 0; i < idx.ns_.size(); ++i) {
-    math::check_range("vector[multi] indexing", name, v_ref.size(), idx.ns_[i]);
-    ret_v.coeffRef(i) = v_ref.coeff(idx.ns_[i] - 1);
-  }
-  return ret_v;
+inline auto rvalue(EigVec&& v, const char* name, const index_multi& idx) {
+  return stan::math::make_holder(
+      [name, &idx](auto& v_ref) {
```

Comment:
In general C++ code yes. In C++ generated by stanc no. This is the same as for any function in math. So as not to litter whole math library with holders we have decided to ignore such cases.

Reply:
I'm not sure I'm following, if stanc generated

```
auto blah = rvalue(rvalue(v, "blah1", index_multi(std::vector<int>{1, 2, 3}), "blah1", index_uni(3)))
```

Wouldn't the `index_multi` fall out of scope here?

Reply:
Nope. It stays in scope until everything in the same line of code is executed.

Reply:
Let me play with this on godbolt for a minute and if I can't get it to break then I'm cool with it

Reply:
Okay so I'm cool with this, as long as we are agreeing that we won't ever use `auto` in transformed data/ transformed parameters in the compiler because then we hit a no-no. But don't we want `auto` in the compiler for the OpenCL stuff? How less performant is it to just make the multi index have a perfect forwarding template and forward the object along into the lambda?

> So as not to litter whole math library with holders we have decided to ignore such cases.

Do you have a link to where that decision was made at? tmk didn't we want to add those?

https://godbolt.org/z/jEYTde3G1

Reply:
> But don't we want auto in the compiler for the OpenCL stuff?

No need. We can directly use `matrix_cl`. As for expressions we can handle them more or less exactly as Eigen ones.

> How less performant is it to just make the multi index have a perfect forwarding template and forward the object along into the lambda?

That would not work if the same index is used multiple times.

> Do you have a link to where that decision was made at? tmk didn't we want to add those?

https://github.com/stan-dev/math/issues/1470#issuecomment-725917281

Reply:
Alright then I'm good with this